### PR TITLE
Improve waiting for the server is in use by the signal that a marker file is created by the build task

### LINKED ISSUE #8956: Flaky test - MSBuildServerTest
### Context

```
System.TimeoutException : Test failed due to timeout: process 18281 is active for more than 30 sec.
```

```
   at Microsoft.Build.UnitTests.Shared.RunnerUtilities.RunProcessAndGetOutput(String process, String parameters, Boolean& successfulExit, Boolean shellExecute, ITestOutputHelper outputHelper) in /Users/runner/work/1/s/src/UnitTests.Shared/RunnerUtilities.cs:line 110
   at Microsoft.Build.UnitTests.Shared.RunnerUtilities.ExecMSBuild(String pathToMsBuildExe, String msbuildParameters, Boolean& successfulExit, Boolean shellExecute, ITestOutputHelper outputHelper) in /Users/runner/work/1/s/src/UnitTests.Shared/RunnerUtilities.cs:line 42
   at Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.MSBuildServerTest() in /Users/runner/work/1/s/src/MSBuild.UnitTests/MSBuildServer_Tests.cs:line 121
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

e.g.: https://dev.azure.com/dnceng-public/public/_build/results?buildId=320229&view=ms.vss-test-web.build-test-results-tab&runId=6566008&resultId=100002&paneView=debug

Looking on the history of the test it seems to fail almost 1 PR a day

### PR DESCRIPTION
**Pull request by GangWang01**
Fixes #8956

### Context
https://github.com/dotnet/msbuild/issues/8956#issuecomment-1614321536

### Changes Made
Improve waiting for the server is in use by the signal that a marker file is created by the build task.

### CODE REVIEW

**Review by rainersigwald — COMMENTED** [review]
It's generally not a good idea to have timeouts like this, or to increase them without detailed data indicating why.

Is it possible to get a distinct signal from the second process that it's going, perhaps a line on stdout or a file that acts as a marker?

**Comment by GangWang01** [comment]
> It's generally not a good idea to have timeouts like this, or to increase them without detailed data indicating why.
> 
> Is it possible to get a distinct signal from the second process that it's going, perhaps a line on stdout or a file that acts as a marker?
It's possible to let the build task create a marker file as the signal.

File: src/MSBuild.UnitTests/MSBuildServer_Tests.cs [hunk 1]
```diff
@@ -185,13 +193,23 @@ public void BuildsWhileBuildIsRunningOnServer()
...
+            watcher.EnableRaisingEvents = true;
             t = Task.Run(() =>
             {
                 RunnerUtilities.ExecMSBuild(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath, sleepProject.Path, out _, false, _output);
             });
 
             // The server will soon be in use; make sure we don't try to use it before that happens.
-            Thread.Sleep(1000);
+            mre.WaitOne();
+            _output.WriteLine("It's OK to go ahead.");
```
Comment (rainersigwald) [inline src/MSBuild.UnitTests/MSBuildServer_Tests.cs:212]: I'd actually move this log line to _before_ the indefinite wait, so if the wait hangs for a very long time we can see that clearly in the log.

File: src/MSBuild.UnitTests/MSBuildServer_Tests.cs [hunk 2]
```diff
@@ -176,7 +180,11 @@ public void BuildsWhileBuildIsRunningOnServer()
         {
             _env.SetEnvironmentVariable("MSBUILDUSESERVER", "1");
             TransientTestFile project = _env.CreateFile("testProject.proj", printPidContents);
-            TransientTestFile sleepProject = _env.CreateFile("napProject.proj", sleepingTaskContents);
+
+            string fileName = "marker.txt";
+            string? dir = Path.GetDirectoryName(project.Path);
+            string markerFilePath = Path.Combine(dir!, fileName);
```
Comment (rainersigwald) [inline src/MSBuild.UnitTests/MSBuildServer_Tests.cs:186]: I would prefer to use `_env.ExpectFile()` here so that the marker gets cleaned up.

**Review by rainersigwald — COMMENTED** [review]
This looks so much better!

File: src/MSBuild.UnitTests/MSBuildServer_Tests.cs (hunk 2 above)
Comment (GangWang01) [reply src/MSBuild.UnitTests/MSBuildServer_Tests.cs]: Thank you! I learned more about `TestEnvironment` from this.

File: src/MSBuild.UnitTests/MSBuildServer_Tests.cs (hunk 1 above)
Comment (GangWang01) [reply src/MSBuild.UnitTests/MSBuildServer_Tests.cs:211]: Instead, I'd like to add one more log before waiting.

**Review by JanKrivanek — APPROVED** [review]
It now looks much more solid then the original test - thanks!

**Pull request merged**